### PR TITLE
Migrate replication controllers to generic etcd

### PR DESCRIPTION
**Pull request by bprashanth**
The following will silently happen as a byproduct of the migration:
- Previously, watchers of the rc would always get an upto date reflection of Status.Replicas (because we listed pods). Now they'll get a 5s stale count (see https://github.com/GoogleCloudPlatform/kubernetes/issues/5745). 
- Update and Create won't explicitly validate anymore, they'll do so through the strategy introduced in https://github.com/GoogleCloudPlatform/kubernetes/pull/5443
  - Validation through generic/etcd returns the wrong error in one known case (https://github.com/GoogleCloudPlatform/kubernetes/issues/5684)
- `List` won't list try to match labels in controller/rest (because I deleted most of the file).
- Controllers no longer need a `PodLister`.
- There are no clients to the controller registry, though there probably will be soon. Everything that cares about controllers hits the REST api. 

Also dropped controller tick to 5s, because if a pod dies it almost always does so within this time

### CODE REVIEW

**Comment by bprashanth** [comment]
Still need to do a little bit of cleanup and write a few more uts, but for the most part I'm convinced this is OK.

@smarterclayton @lavalamp

**Comment by mbforbes** [comment]
@lavalamp is out until Wednesday (I believe) and I feel bad explicitly giving work to @smarterclayton (as he's in a different org), so temporarily assigning to @davidopp; feel free to redistribute.

**Comment by smarterclayton** [comment]
I'll review - I'm in that part of the codebase right now anyway.  Don't few bad assigning things to me, but it's somewhat like assigning to Tim or Brian in terms of our ability to review :)

**Comment by derekwaynecarr** [comment]
I am happy to review as well.  I tend to have more time than Clayton and also know this area well :-)

**Comment by davidopp** [comment]
Clayton or Derek, if you want to review this please assign it to yourself, otherwise I will review i tthis week.

File: pkg/registry/controller/etcd/etcd.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+		},
+		// Used to match objects based on labels/fields for list and watch
+		PredicateFunc: func(label labels.Selector, field fields.Selector) generic.Matcher {
+			return controller.MatchController(label, field)
+		},
+		EndpointName: "replicationControllers",
+
+		Helper: h,
+	}
+	store.CreateStrategy = controller.Strategy
```
Comment (smarterclayton) [inline pkg/registry/controller/etcd/etcd.go]: Move these to struct init fields

File: pkg/registry/controller/etcd/etcd_test.go [hunk 2]
```diff
@@ -0,0 +1,594 @@
...
+	case <-watching.ResultChan():
+		t.Error("unexpected result from result channel")
+	case <-time.After(time.Millisecond * 100):
+		// expected case
+	}
+}
+
+func TestCreate(t *testing.T) {
+	storage, fakeClient := newStorage(t)
+	test := resttest.New(t, storage, fakeClient.SetError)
```
Comment (smarterclayton) [inline pkg/registry/controller/etcd/etcd_test.go]: Can you also add in the resttest.TestDelete tests as well?

**Comment by smarterclayton** [comment]
Two minor comments

File: pkg/registry/controller/etcd/etcd.go (hunk 1 above)
Comment (bprashanth) [inline pkg/registry/controller/etcd/etcd.go]: I'm going to add a subresource for /status, but I guess I'll just do that in the next cl.

Comment (smarterclayton) [inline pkg/registry/controller/etcd/etcd.go]: Ok, that's fine then.  Leave it.

----- Original Message -----

File: pkg/registry/controller/etcd/etcd_test.go (hunk 2 above)
Comment (bprashanth) [inline pkg/registry/controller/etcd/etcd_test.go]: Done, also modified pod/etcd_test to use resttest.TestDelete (which calls into TestNoGraceful). The delete test is somewhat basic, though.

File: pkg/registry/controller/etcd/etcd.go (hunk 1 above)
Comment (bprashanth) [inline pkg/registry/controller/etcd/etcd.go]: Changed anyway, things might happen that prevent that cl from going in so it's cleaner this way.

**Comment by bprashanth** [comment]
PTAL

**Comment by smarterclayton** [comment]
LGTM

**Comment by bprashanth** [comment]
Haven't run e2e post rebase yet

**Comment by bprashanth** [comment]
e2e done, this should be good to go

**Pull request merged**